### ISSUE TITLE
Integrating SOFA with Apollo Federation

### ISSUE THREAD

**Issue by John-Blaine-Jr** [Urigo/SOFA#214]
I have a microservice API that is using SOFA and seems to be serving up data via REST just fine: 
`
const { ApolloServer } = require('apollo-server-express');
const { buildFederatedSchema } = require('@apollo/federation');
const express = require('express');
const { useSofa } = require('sofa-api');

const { resolvers } = require('./schema/resolvers');
const { typeDefs } = require('./schema/typeDefs');

const graphQL = new ApolloServer({
  schema: buildFederatedSchema(
    {
      typeDefs,
      resolvers
    }
  )
});

const app = express();

app.use('/api', useSofa({
  schema: buildFederatedSchema({
    typeDefs,
    resolvers
  })
}));

graphQL.applyMiddleware({ app });

const port = process.env.PORT || 4001;

app.listen({ port }, () => {
  console.log(`🚀  Server ready http://localhost:${port}`);
});
`
But then, when I try to federate that microservice, and any number of microservices, into a gateway API as instantiated below, I recieve a `null` value whenever I make a REST query.
`
const express = require('express');
const { useSofa, OpenAPI } = require('sofa-api');
const { ApolloServer } = require('apollo-server-express');
const { ApolloGateway } = require("@apollo/gateway");

const gateway = new ApolloGateway({
  serviceList: [
    { name: "salesLineItem", url: `${process.env.NUTRIEN_SALES_DOCUMENT_SERVICE_URL}/graphql` },
    { name: "salesDocument", url: `${process.env.NUTRIEN_SALES_LINE_ITEM_SERVICE_URL}/graphql` },
    { name: "productPrice", url: `${process.env.NUTRIEN_PRODUCT_PRICE_SERVICE_URL}/graphql` },
  ]
});

(async () => {
  const { schema, executor } = await gateway.load();
  console.log('Gateway schema', JSON.stringify(schema));
  const graphql = new ApolloServer({ schema, executor });

  const app = express();

  app.use(
    '/api',
    useSofa({
      schema
    }),
  );

  graphql.applyMiddleware({ app });
  
  app.listen({ port: process.env.PORT }, () => {
    console.log(`🚀  Server ready http://localhost:${process.env.PORT}`);
  });
})();
`
So, the same query succeeds when made against the microservice, but returns `null` Is it not possible to use SOFA with a federated Apollo server?

**Comment by ardatan** [comment]
Could you try the following?
```ts
useSofa({
  schema,
  execute: ({ source, variableNames, operationName }) => apolloServer.executeOperation({ query: source, variables: variableNames, operationName })
})
```

**Comment by John-Blaine-Jr** [comment]
I think you meant:
```
useSofa({
  schema,
  executor /* instead of execute */: ({ source, variableNames, operationName }) => apolloServer.executeOperation({ query: source, variables: variableNames, operationName })
})
```

**Comment by ardatan** [comment]
@John-Blaine-Jr I meant `execute` not `executor`.

**Comment by ardatan** [comment]
@John-Blaine-Jr Any news?

**Comment by John-Blaine-Jr** [comment]
@ardatan No, I tried using `execute` instead of `executor` and was getting an error instead of `null`. I'll send the error when I have some time later

**Comment by John-Blaine-Jr** [comment]
@ardatan Sorry for the wait, busy at work. Here is the error message I am receiving when I try to implement your suggestion using`execute`:
```{
    "message": "Cannot convert undefined or null to object",
    "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
    }
}```

**Comment by byteninja** [comment]
```
useSofa({
  schema,
  execute: ({ source, variableValues, operationName }) => apolloServer.executeOperation({ query: source, variables: variableValues, operationName })
})
```
is the right code i think.
this works fine for me on queries, but fails for mutations where the variables should be parsed from the body. The variableValues are always an empty Object ({}) in this case and the mutation fails

**Comment by yuraxdrumz** [comment]
Don't know if its too late @byteninja, but I stumbled on this issue and found the following:

1. `variableValues` come from query params on GET requests and body on POST requests
2. inner `executeQueryPlan` in apollo gateway looks like so

```
        try {
          let postProcessingErrors: GraphQLError[];
          const variables = requestContext.request.variables;
          ({ data, errors: postProcessingErrors } = computeResponse({
            operation,
            variables,
            input: unfilteredData,
            introspectionHandling: (f) => executeIntrospection(
              operationContext.schema,
              f.expandFragments().toSelectionNode(),
              operationContext.operation.variableDefinitions,
              variables,
            ),
          }));
```

Where it expects the variables from the context, but the issue is 
`contextValue.request.variables` is empty on POST

The solution I found that works is setting `contextValue.request.variables = variableValues;` inside the custom `execute` function

Example

```
      execute: ({ document, variableValues, operationName, contextValue }) => {
        contextValue.request.variables = variableValues;
        return executor({ queryHash: operationName, operationName: operationName, document: document, request: contextValue.request });
      },
```

Seems good so far
